pkg/networkservice/common/dial/client.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dial will dial up a grpc.ClientConnInterface if a client *url.URL is provided in the ctx, retrievable by
+// clienturlctx.ClientURL(ctx) and put the resulting grpc.ClientConnInterface into the ctx using clientconn.Store(..)
+// where it can be retrieved by other chain elements using clientconn.Load(...)
+package dial
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clientconn"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/utils/metadata"
+	"github.com/networkservicemesh/sdk/pkg/tools/clienturlctx"
+	"github.com/networkservicemesh/sdk/pkg/tools/clock"
+	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
+)
+
+type dialClient struct {
+	chainCtx    context.Context
+	dialOptions []grpc.DialOption
+	dialTimeout time.Duration
+}
+
+// NewClient - returns new dial chain element
+func NewClient(chainCtx context.Context, opts ...Option) networkservice.NetworkServiceClient {
+	o := &option{}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return &dialClient{
+		chainCtx:    chainCtx,
+		dialOptions: o.dialOptions,
+		dialTimeout: o.dialTimeout,
+	}
+}
+
+func (d *dialClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
```

Comment:
We have a problem in the existing connect which is related to SVIDs rotation over time - if we have a gRPC connection that outlives such rotation, we have different token and SSL certificates in Request and so it fails in `authorize`.
It is partially solved by [`resetting`]( https://github.com/networkservicemesh/sdk/blob/main/pkg/tools/resetting/transport_credentials.go), but there is still some chance for this issue to happen and it is big enough to fail NSE Composition example.
Another issue, that isn't any covered by resetting happens when spire agent restarts - https://github.com/networkservicemesh/deployments-k8s/issues/2393.
The solution we currently want to use is to create per-Request gRPC connections - in such case even if we meet issue when we have different spire certificates in token and in peer, it will definitely be solved with the subsequent Request.
I already have a PR for this (https://github.com/networkservicemesh/sdk/pull/1069), so probably we want either to merge my PR and rebase this one, or just fix this one with the corresponding changes.

Reply:
I get this... the issue I'm trying to figure out is what to do with monitor in that case.  Monitor needs a grpc.ClientConnInterface that is persistent between calls... open to ideas and suggestions.

Reply:
I currently use the following approach in healing (it also need monitor connection):
1. Request establishes gRPC connection.
2. Monitor client starts on it.
3. Subsequent Request/Close closes existing monitor client and establishes new gRPC connection.

If remote side fails before [3], monitor receives failure. If remote side fails during [3], Request returns an error.

So here you cannot close monitor client by closing client context, because there is no client recreation on every Request. But you can possibly just close gRPC connection, and so monitor client can additionally check if connection is broken or if it was closed and there is no need to trigger an event, so it should be almost the same.

Reply:
@Bolodya1997
> Subsequent Request/Close closes existing monitor client and establishes new gRPC connection.

This is the approach I also adopted :)